src/operator/recursive.rs

```diff
@@ -0,0 +1,432 @@
+    /// elements in the Z-set no longer changes. Hence, strictly speaking
+    /// this circuit computes the fixed point of equation
+    /// `x' = distinct(f(i+Δi, x'))`.
+    ///
+    /// Finally, the `δ0` block in the diagram represents the
+    /// [`delta0`](`crate::circuit::Stream::delta0`) operator, which imports
+    /// streams from the parent circuit into the nested circuit.  This
+    /// operator must be instantiated manually by the closure `f` for each
+    /// input stream.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use dbsp::{
+    ///     circuit::{Root, Stream},
+    ///     operator::Generator,
+    ///     time::NestedTimestamp32,
+    ///     trace::ord::OrdZSet,
+    ///     zset, zset_set,
+    /// };
+    /// use std::vec;
+    ///
+    /// // Propagate labels along graph edges.
+    /// let root = Root::build(move |circuit| {
+    ///     // Graph topology.
+    ///     let mut edges: vec::IntoIter<OrdZSet<(usize, usize), isize>> = vec![
+    ///         // Start with four nodes connected in a cycle.
+    ///         zset_set! { (1, 2), (2, 3), (3, 4), (4, 1) },
+    ///         // Add an edge.
+    ///         zset_set! { (4, 5) },
+    ///         // Remove an edge, breaking the cycle.
+    ///         zset! { (1, 2) => -1 },
+    ///     ]
+    ///     .into_iter();
+    ///
+    ///     let edges: Stream<_, OrdZSet<(usize, usize), isize>> =
+    ///         circuit
+    ///             .add_source(Generator::new(move || edges.next().unwrap()));
+    ///
+    ///     // Initial labeling of the graph.
+    ///     let mut init_labels: vec::IntoIter<OrdZSet<(usize, &'static str), isize>> = vec![
+    ///         // Start with a single label on node 1.
+    ///         zset_set! { (1, "l1") },
+    ///         // Add a label to node 2.
+    ///         zset_set! { (2, "l2") },
+    ///         zset! { },
+    ///     ]
+    ///     .into_iter();
+    ///
+    ///     let init_labels: Stream<_, OrdZSet<(usize, &'static str), isize>> =
+    ///         circuit
+    ///             .add_source(Generator::new(move || init_labels.next().unwrap()));
+    ///
+    ///     // Expected _changes_ to the output graph labeling after each clock cycle.
+    ///     let mut expected_outputs: vec::IntoIter<OrdZSet<(usize, &'static str), isize>> = vec![
+    ///         zset! { (1, "l1") => 1, (2, "l1") => 1, (3, "l1") => 1, (4, "l1") => 1 },
+    ///         zset! { (1, "l2") => 1, (2, "l2") => 1, (3, "l2") => 1, (4, "l2") => 1, (5, "l1") => 1, (5, "l2") => 1 },
+    ///         zset! { (2, "l1") => -1, (3, "l1") => -1, (4, "l1") => -1, (5, "l1") => -1 },
+    ///     ]
+    ///     .into_iter();
+    ///
+    ///     let labels = circuit.recursive(|child, labels: Stream<_, OrdZSet<(usize, &'static str), isize>>| {
+    ///         // Import `edges` and `init_labels` relations from the parent circuit.
+    ///         let edges = edges.delta0(child);
```

Comment:
You mean `delta0`? If you don't use it, you can't access an external stream, so it's hard to forget.  It might be possible to automate this step as well, but I was not able to think of a nice way.